File: src/main/java/no/nav/syfo/oidc/OIDCUtil.java
package no.nav.syfo.oidc;

import no.nav.security.token.support.core.context.TokenValidationContext;
import no.nav.security.token.support.core.context.TokenValidationContextHolder;
import no.nav.syfo.service.ws.OnBehalfOfOutInterceptor;
import org.apache.cxf.endpoint.Client;


public class OIDCUtil {

    public static void leggTilOnBehalfOfOutInterceptorForOIDC(Client client, String OIDCToken) {
        client.getRequestContext().put(OnBehalfOfOutInterceptor.REQUEST_CONTEXT_ONBEHALFOF_TOKEN_TYPE, OnBehalfOfOutInterceptor.TokenType.OIDC);
        client.getRequestContext().put(OnBehalfOfOutInterceptor.REQUEST_CONTEXT_ONBEHALFOF_TOKEN, OIDCToken);
    }

    public static String getSubjectEksternMedThrows(TokenValidationContextHolder contextHolder) {
        try {
            TokenValidationContext context = contextHolder.getTokenValidationContext();
            return context.getClaims(OIDCIssuer.EKSTERN).getSubject();
        } catch (NullPointerException e) {
            throw new RuntimeException("Fant ikke subject for OIDCIssuer Ekstern");
        }
    }

    public static String getIssuerToken(TokenValidationContextHolder contextHolder, String issuer) {
        TokenValidationContext context = contextHolder.getTokenValidationContext();
        return context.getJwtToken(issuer).getTokenAsString();
    }
}

File: src/main/kotlin/no/nav/syfo/oidc/OIDCUtil.kt
package no.nav.syfo.oidc

import no.nav.security.token.support.core.context.TokenValidationContextHolder
import no.nav.security.token.support.core.jwt.JwtTokenClaims
import no.nav.syfo.oidc.OIDCIssuer.EKSTERN
import no.nav.syfo.tokenx.TokenXUtil.TokenXIssuer.TOKENX
import java.text.ParseException

object OIDCUtil {

    private val PID_CLAIM = "pid"

    @JvmStatic
    fun getSubjectEksternMedThrows(contextHolder: TokenValidationContextHolder): String {
        val issuer = EKSTERN
        val context = contextHolder.tokenValidationContext
        return context.getClaims(issuer)?.let {
            getPidClaimFromClaimsSet(it, issuer)
        } ?: throw RuntimeException("Fant ikke 'subject'-claim i token fra issuer: $issuer")
    }

    @JvmStatic
    fun getIssuerToken(contextHolder: TokenValidationContextHolder, issuer: String): String {
        val context = contextHolder.tokenValidationContext
        return context.getJwtToken(issuer)?.tokenAsString ?: throw RuntimeException("Klarte ikke hente token fra issuer: $issuer")
    }

    @JvmStatic
    fun getSluttbrukerToken(contextHolder: TokenValidationContextHolder): String {
        val context = contextHolder.tokenValidationContext
        val eksternToken = context.getJwtToken(EKSTERN)
        val tokenxToken = context.getJwtToken(TOKENX)
        return when {
            eksternToken != null ->
                eksternToken.tokenAsString
            tokenxToken != null ->
                tokenxToken.tokenAsString
            else ->
                throw RuntimeException("Klarte ikke hente token fra issuer $EKSTERN eller $TOKENX")
        }
    }

    @JvmStatic
    fun getPidClaimFromClaimsSet(jwtTokenClaims: JwtTokenClaims, issuer: String): String {
        return try {
            jwtTokenClaims.getStringClaim(PID_CLAIM)
        } catch (e: ParseException) {
            throw RuntimeException("Fant ikke pid-claim i tokenet fra $issuer")
        }
    }

}